### PR TITLE
Fix checkpoint loading for hybrid checkpoint

### PR DESCRIPTION
**Pull request by nvcsathe**
- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by guihong-nv — COMMENTED** [review]
One blocking regression in the checkpoint layout selection path.

File: megatron/training/checkpointing.py [hunk 1]
```diff
@@ -2130,19 +2135,25 @@ def load_checkpoint(
...
 
     # Both model-space torch_dist and fsdp_dtensor checkpoints carry model-keyed
     # optimizer state that can be retargeted from GPTModel to HybridModel.
     gpt_compat_layer_maps, gpt_compat_load_optim = (
         _maybe_setup_gpt_to_hybrid_load(args, ckpt_args, model)
-        if ckpt_format in ('torch_dist', 'fsdp_dtensor') and state_dict is not None
+        if (
+            ckpt_format in ('torch_dist', 'fsdp_dtensor')
+            and state_dict is not None
+            and checkpoint_args_available
```
Comment (guihong-nv) [inline megatron/training/checkpointing.py]: [P1] Preserve GPT→Hybrid loading without saved args

Requiring `checkpoint_args_available` here makes every checkpoint with missing or `None` `args` use the native HybridModel layout. That fixes no-args hybrid checkpoints, but regresses no-args GPT checkpoints: the previous code passed an empty namespace to `_maybe_setup_gpt_to_hybrid_load`, which triggered the required key retargeting. Distributed checkpoints without a cached args blob are supported elsewhere. Without retargeting, the loader requests HybridModel keys from a homogeneous GPT checkpoint, causing strict-load failures or partial fresh initialization under permissive loading.

Please infer the architecture from checkpoint key metadata or require an explicit layout hint, and add end-to-end coverage for both no-args GPT and no-args hybrid checkpoints. The new helper-only tests do not exercise either `load_checkpoint` path.

File: megatron/training/checkpointing.py [hunk 2]
```diff
@@ -2010,7 +2016,20 @@ def _contains_hybrid_model(module):
     ckpt_pattern = getattr(ckpt_args, 'hybrid_layer_pattern', None) or getattr(
         ckpt_args, 'hybrid_override_pattern', None
     )
-    if runtime_is_hybrid == bool(ckpt_pattern):
+    checkpoint_is_hybrid = bool(ckpt_pattern)
+    if runtime_is_hybrid and ckpt_args is None:
```
Comment (guihong-nv) [inline megatron/training/checkpointing.py]: [P2] Infer a no-args checkpoint layout independently of the runtime type

This guard means tensor metadata is inspected only when the destination runtime is a `HybridModel`. If `ckpt_args is None` and a HybridModel checkpoint is loaded into a non-hybrid runtime, `checkpoint_is_hybrid` remains `False`; the equality check below therefore incorrectly concludes that the layouts match and bypasses the intended reverse-mismatch `RuntimeError`. Under permissive loading this can also permit a partial load of overlapping state.

Please run the metadata inference whenever `ckpt_args is None`, then compare the inferred checkpoint type with `runtime_is_hybrid`. An inverse integration test for a no-args HybridModel checkpoint loaded into a GPT runtime would cover this branch.

**Review by guihong-nv — COMMENTED** [review]
One additional correctness issue in the no-args checkpoint layout inference.

**Review by guihong-nv — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]

**Review by shanmugamr1992 — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]

**Comment by sidsingh-nvidia** [comment]
/ok to test [20cb5ae](https://github.com/NVIDIA/Megatron-LM/pull/6107/commits/20cb5ae3096a99670d646c0a0ba6e71e8c4f0fc5)

**Comment by sidsingh-nvidia** [comment]
/ok to test [3114124](https://github.com/NVIDIA/Megatron-LM/pull/6107/commits/311412416ee45ccecdc815506c9efa7ff30c9c58)

**Comment by sidsingh-nvidia** [comment]
/ok to test [d356927](https://github.com/NVIDIA/Megatron-LM/pull/6107/commits/d356927415f6f98e6b3f779d15481a2a5f4068ce)
